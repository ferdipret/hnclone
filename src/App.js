import React, { Component } from 'react';

class App extends Component {
  render() {
    return (
      <div className="App">
        HackerNews App
      </div>
    );
  }
}

export default App;
