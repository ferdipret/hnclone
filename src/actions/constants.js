export const SET_STORIES_IDS_LIST = 'SET_STORIES_IDS_LIST'
export const SET_STORIES_LIST_DATA = 'SET_STORIES_LIST_DATA'
export const API_ENTRY_POINT = 'https://hacker-news.firebaseio.com/v0'
