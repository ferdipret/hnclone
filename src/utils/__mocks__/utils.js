const n = 500

// Will create an array of (0 [, ...n])
export const topStoriesIds = Array.apply(null, { length: n }).map(
  Number.call,
  Number,
)

export const defaultIds = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

export const calculatedIds = [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]
