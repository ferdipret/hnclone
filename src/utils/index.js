export function calculateNextStoryIds(idList, current) {
  return idList.slice(current, current + 10)
}

export function cleanUrl(string) {
  const regex = /\/\/([^\/,\s]+?\.[^\/,\s]+?)(?=\/|,|\s|$|\?|#)/g
  const cleanArray = regex.exec(string)

  if (cleanArray) {
    return cleanArray[1].replace(/^www\./g, '')
  }
}
