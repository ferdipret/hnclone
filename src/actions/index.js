import * as constants from './constants'

export function setStoriesIdsList(payload) {
  return {
    type: constants.SET_STORIES_IDS_LIST,
    payload,
  }
}

export function setStoriesListData(payload) {
  return {
    type: constants.SET_STORIES_LIST_DATA,
    payload,
  }
}
