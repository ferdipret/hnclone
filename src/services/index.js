import axios from 'axios'
import { API_ENTRY_POINT } from '../actions/constants'

export const fetchTopStories = endpoint => {
  const url = API_ENTRY_POINT + endpoint

  return axios.get(url)
}
