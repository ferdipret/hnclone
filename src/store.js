import * as constants from './actions/constants'

export const defaultState = {
  storiesIdsList: [],
  stories: [],
  numberOfStoriesLoaded: 0,
}

export default (state = defaultState, action) => {
  let newState

  switch (action.type) {
    case constants.SET_STORIES_IDS_LIST:
      newState = {
        ...state,
        storiesIdsList: action.payload,
      }
      break

    case constants.SET_STORIES_LIST_DATA:
      newState = {
        ...state,
        stories: [...state.stories, ...action.payload],
      }
      break

    default:
      newState = state
      break
  }

  return newState
}
