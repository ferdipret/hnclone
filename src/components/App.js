import React, { Component } from 'react'
import store from '../store'

class App extends Component {
  constructor() {
    super()

    this.state = store(undefined, {})
  }

  dispatch = action => {
    action &&
      action.type &&
      this.setState(prevState => store(prevState, action))
  }

  render() {
    return <h1 className="app-title">App Component</h1>
  }
}

export default App
