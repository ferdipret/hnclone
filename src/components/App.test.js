import React from 'react'
import { mount, shallow } from 'enzyme'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import App from './App'
import store, { defaultState } from '../store'
import { setStoriesIdsList } from '../actions'
import { fetchTopStories } from '../services'
import * as constants from '../actions/constants'
import { calculateNextStoryIds } from '../utils'

const mock = new MockAdapter(axios, { delayResponse: 2000 })

export const responseData = ['1', '2', '3']

mock
  .onGet(`${constants.API_ENTRY_POINT}/topstories.json`)
  .reply(200, responseData)

describe('App', () => {
  const app = shallow(<App />)

  it('renders properly', () => {
    expect(app).toMatchSnapshot()
  })

  it('should get it`s default state from the `store` function', () => {
    expect(app.state()).toEqual(defaultState)
  })

  describe('calling `dispatch`', () => {
    afterEach(() => {
      app.setState(defaultState)
    })

    it('should dispatch actions to the store and update the `App` state', () => {
      const expectedState = ['1', '2', '3']
      const action = setStoriesIdsList(expectedState)

      app.instance().dispatch(action)
      expect(app.state().storiesIdsList).toEqual(expectedState)
    })

    it('should not change state if there is no valid action passed in as an argument', () => {
      const action = {}

      app.instance().dispatch(action)
      expect(app.state()).toEqual(defaultState)
    })
  })

  describe('when mounted', () => {
    it('should call the fetchTopStories service', () => {
      return fetchTopStories('/topstories').then(() => {
        expect(app.state('storiesIdsList')).toEqual(['1', '2', '3'])
      })
    })
  })

  describe('when calling `fetchMoreStories`', () => {
    it('should fetch the next 10 stories', () => {
      app.instance().fetchMoreStories(0)

      // expect(app.state('stories')).toEqual()
    })
  })
})
