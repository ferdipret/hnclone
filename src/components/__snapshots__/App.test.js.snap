// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h1
          className="app-title"
>
          App Component
</h1>,
        <StoriesList
          stories={Array []}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "App Component",
          "className": "app-title",
        },
        "ref": null,
        "rendered": "App Component",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "stories": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h1
            className="app-title"
>
            App Component
</h1>,
          <StoriesList
            stories={Array []}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "App Component",
            "className": "app-title",
          },
          "ref": null,
          "rendered": "App Component",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "stories": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
